### ISSUE TITLE
[BUG] unable to provision node driver clusters 

### ISSUE THREAD

**Issue by slickwarren** [rancher/rancher#42972]
**Rancher Server Setup**
- Rancher version: v2.8-head (e5ce94a)
- Installation option (Docker install/Helm Chart):
   - If Helm Chart, Kubernetes Cluster and version (RKE1, RKE2, k3s, EKS, etc): rke1
- Proxy/Cert Details:

**Information about the Cluster**
- Kubernetes version: any (tested on 1.27 and 1.26)
- Cluster Type (Local/Downstream): downstream, rke2, rke1
   - If downstream, what type of cluster? (Custom/Imported or specify provider for Hosted/Infrastructure Provider): aws

**User Information**
- What is the role of the user logged in? (Admin/Cluster Owner/Cluster Member/Project Owner/Project Member/Custom)
  - If custom, define the set of permissions: tested with standard user, cluster owner and as admin

**Describe the bug**

user is hit with an error when creating a machine in the provider when creating a vanilla cluster

**To Reproduce**

* provision an aws cluster with 1 node 

**Result**
node doesn't ever get properly provisioned (see screenshot)

**Screenshots**

![Screen Shot 2023-09-27 at 11 00 27 AM](https://github.com/rancher/rancher/assets/16691014/22d073c1-7ad8-4218-bb82-6a53bb12dbd9)

**Additional context**

**Comment by Sahota1225** [comment]
@slickwarren - can you validate this issue with alpha 2.8.0 -alpha2 ?

**Comment by sennerholm** [comment]
We got simular "problem" with the 2.7.7 that changed to use rancher/machine:v0.15.0-rancher104 that are used in Rancher 2.7.7(added after 2.7.7rc5). We got the following in the pod logs:

```
kubectl logs -f alpha-lab10321-control-plane-8f0208a4-4zj4d-machine-provisk58m6
Running pre-create checks...
(alpha-lab10321-control-plane-8f0208a4-4zj4d) Connecting to vSphere for pre-create checks...
(alpha-lab10321-control-plane-8f0208a4-4zj4d) Using datacenter /MyDatacenter
Error with pre-create check: "unexpected EOF"
The default lines below are for a sh/bash shell, you can specify the shell you're using, with the --shell flag.
```

**Comment by sennerholm** [comment]
On 2.7.7 we get futher if we override the version to 105, by setting environment for the rancher deployment
```

            - name: CATTLE_MACHINE_PROVISION_IMAGE
              value: 'rancher/machine:v0.15.0-rancher105'
```

**Comment by snasovich** [comment]
@sennerholm , thank you for communicating a workaround.
Couple notes:
* Recommend setting the version to `rancher/machine:v0.15.0-rancher100` as the changes between 101 and 105 were not intended for 2.7.7 and got there inadvertently. Most likely we will be reverting to that version in the next 2.7.x release.
* Setting `CATTLE_MACHINE_PROVISION_IMAGE` only applies to RKE2 provisioned clusters as far as I remember. RKE1 provisioned clusters will still use the version Rancher shipped with.

**Comment by susesgartner** [comment]
Ticket #42972 - Test Results - ✅ 
---
Verified with HA Helm on Rancher v2.8-bfbf05a582d8a263a7385016f00f4215e6bcb876-head:

1. Create a 1 node downstream AWS cluster 

Result: Cluster gets provisioned as expected with no issues.

**Comment by sennerholm** [comment]
> @sennerholm , thank you for communicating a workaround. Couple notes:
> 
> * Recommend setting the version to `rancher/machine:v0.15.0-rancher100` as the changes between 101 and 105 were not intended for 2.7.7 and got there inadvertently. Most likely we will be reverting to that version in the next 2.7.x release.

I tried downgrading first, but it then it logs complains about wrong arguments. We are deploying to VMware.

Sincerely
Mikael

**Comment by snasovich** [comment]
@sennerholm , thank you. Could you please share your findings from testing on version 100 (including any warnings you may be getting)?

**Comment by sennerholm** [comment]
> @sennerholm , thank you. Could you please share your findings from testing on version 100 (including any warnings you may be getting)?

Will try to get back tomorrow, Monday. But it was some argument missing och similar in the machine logs

**Comment by sennerholm** [comment]
Only tried once and then rolled forward to 105 version, could have a temporary failure, couldn't found anything else in 2.7.7 that should have changed arguments since rc5 I tested.

**Comment by sennerholm** [comment]
So I got an error with 100
`flag provided but not defined: -vmwarevsphere-graceful-shutdown-timeout`

I started with 105, changing environment variable to point to 100 version, waited for rancher deployment to update, after that changed nr of worker in pool from 3 to 4 and got this from the pod:

```
 kubectl logs -f alpha-lab10321-worker1-e88fa4ef-2x4cd-machine-provision-m4wvs
Incorrect Usage.

Usage: docker-machine create [OPTIONS] [arg...]

Create a machine

Description:
   Run 'rancher-machine create --driver name --help' to include the create flags for that driver in the help text.

Options:

   --custom-install-script                                                                                              Use a custom provisioning script instead of installing docker
   --driver, -d "virtualbox"                                                                                            Driver to create machine with. [$MACHINE_DRIVER]
   --engine-env [--engine-env option --engine-env option]                                                               Specify environment variables to set in the engine
   --engine-insecure-registry [--engine-insecure-registry option --engine-insecure-registry option]                     Specify insecure registries to allow with the created engine
   --engine-install-url "https://get.docker.com"                                                                        Custom URL to use for engine installation [$MACHINE_DOCKER_INSTALL_URL]
   --engine-label [--engine-label option --engine-label option]                                                         Specify labels for the created engine
   --engine-opt [--engine-opt option --engine-opt option]                                                               Specify arbitrary flags to include with the created engine in the form flag=value
   --engine-registry-mirror [--engine-registry-mirror option --engine-registry-mirror option]                           Specify registry mirrors to use [$ENGINE_REGISTRY_MIRROR]
   --engine-storage-driver                                                                                              Specify a storage driver to use with the engine
   --hostname-override                                                                                                  Specify hostname to use during cloud-init instead of default generated hostname
   --swarm                                                                                                              Configure Machine to join a Swarm cluster
   --swarm-addr                                                                                                         addr to advertise for Swarm (default: detect and use the machine IP)
   --swarm-discovery                                                                                                    Discovery service to use with Swarm
   --swarm-experimental                                                                                                 Enable Swarm experimental features
   --swarm-host "tcp://0.0.0.0:3376"                                                                                    ip/socket to listen on for Swarm master
   --swarm-image "swarm:latest"                                                                                         Specify Docker image to use for Swarm [$MACHINE_SWARM_IMAGE]
   --swarm-join-opt [--swarm-join-opt option --swarm-join-opt option]                                                   Define arbitrary flags for Swarm join
   --swarm-master                                                                                                       Configure Machine to be a Swarm master
   --swarm-opt [--swarm-opt option --swarm-opt option]                                                                  Define arbitrary flags for Swarm master
   --swarm-strategy "spread"                                                                                            Define a default scheduling strategy for Swarm
   --tls-san [--tls-san option --tls-san option]                                                                        Support extra SANs for TLS certs
   --vmwarevsphere-boot2docker-url                                                                                      vSphere URL for boot2docker image [$VSPHERE_BOOT2DOCKER_URL]
   --vmwarevsphere-cfgparam [--vmwarevsphere-cfgparam option --vmwarevsphere-cfgparam option]                           vSphere vm configuration parameters (used for guestinfo) [$VSPHERE_CFGPARAM]
   --vmwarevsphere-clone-from                                                                                           If you choose creation type clone a name of what you want to clone is required [$VSPHERE_CLONE_FROM]
   --vmwarevsphere-cloud-config                                                                                         Filepath to a cloud-config yaml file to put into the ISO user-data [$VSPHERE_CLOUD_CONFIG]
   --vmwarevsphere-cloudinit                                                                                            vSphere cloud-init filepath or url to add to guestinfo, filepath will be read and base64 encoded before adding [$VSPHERE_CLOUDINIT]
   --vmwarevsphere-content-library                                                                                      If you choose to clone from a content library template specify the name of the library [$VSPHERE_CONTENT_LIBRARY]
   --vmwarevsphere-cpu-count "2"                                                                                        vSphere CPU number for docker VM [$VSPHERE_CPU_COUNT]
   --vmwarevsphere-creation-type "legacy"                                                                               Creation type when creating a new virtual machine. Supported values: vm, template, library, legacy [$VSPHERE_CREATION_TYPE]
   --vmwarevsphere-custom-attribute [--vmwarevsphere-custom-attribute option --vmwarevsphere-custom-attribute option]   vSphere custom attribute, format key/value e.g. '200=my custom value'
   --vmwarevsphere-datacenter                                                                                           vSphere datacenter for virtual machine [$VSPHERE_DATACENTER]
   --vmwarevsphere-datastore                                                                                            vSphere datastore for virtual machine [$VSPHERE_DATASTORE]
   --vmwarevsphere-datastore-cluster                                                                                    vSphere datastore cluster for virtual machine [$VSPHERE_DATASTORE_CLUSTER]
   --vmwarevsphere-disk-size "20480"                                                                                    vSphere size of disk for docker VM (in MB) [$VSPHERE_DISK_SIZE]
   --vmwarevsphere-folder                                                                                               vSphere folder for the docker VM. This folder must already exist in the datacenter [$VSPHERE_FOLDER]
   --vmwarevsphere-hostsystem                                                                                           vSphere compute resource where the docker VM will be instantiated. This can be omitted if using a cluster with DRS [$VSPHERE_HOSTSYSTEM]
   --vmwarevsphere-memory-size "2048"                                                                                   vSphere size of memory for docker VM (in MB) [$VSPHERE_MEMORY_SIZE]
   --vmwarevsphere-network [--vmwarevsphere-network option --vmwarevsphere-network option]                              vSphere network where the virtual machine will be attached [$VSPHERE_NETWORK]
   --vmwarevsphere-os "linux"                                                                                           If using a non-B2D image you can specify the desired machine OS
   --vmwarevsphere-password                                                                                             vSphere password [$VSPHERE_PASSWORD]
   --vmwarevsphere-pool                                                                                                 vSphere resource pool for docker VM [$VSPHERE_POOL]
   --vmwarevsphere-ssh-password "tcuser"                                                                                If using a non-B2D image you can specify the ssh password [$VSPHERE_SSH_PASSWORD]
   --vmwarevsphere-ssh-port "22"                                                                                        If using a non-B2D image you can specify the ssh port [$VSPHERE_SSH_PORT]
   --vmwarevsphere-ssh-user "docker"                                                                                    If using a non-B2D image you can specify the ssh user [$VSPHERE_SSH_USER]
   --vmwarevsphere-ssh-user-group "staff"                                                                               If using a non-B2D image the uploaded keys will need chown'ed, defaults to staff e.g. docker:staff [$VSPHERE_SSH_USER_GROUP]
   --vmwarevsphere-tag [--vmwarevsphere-tag option --vmwarevsphere-tag option]                                          vSphere tag id e.g. urn:xxx
   --vmwarevsphere-username                                                                                             vSphere username [$VSPHERE_USERNAME]
   --vmwarevsphere-vapp-ipallocationpolicy                                                                              vSphere vApp IP allocation policy. Supported values are: dhcp, fixed, transient and fixedAllocated [$VSPHERE_VAPP_IPALLOCATIONPOLICY]
   --vmwarevsphere-vapp-ipprotocol                                                                                      vSphere vApp IP protocol for this deployment. Supported values are: IPv4 and IPv6 [$VSPHERE_VAPP_IPPROTOCOL]
   --vmwarevsphere-vapp-property [--vmwarevsphere-vapp-property option --vmwarevsphere-vapp-property option]            vSphere vApp properties [$VSPHERE_VAPP_PROPERTY]
   --vmwarevsphere-vapp-transport                                                                                       vSphere OVF environment transports to use for properties. Supported values are: iso and com.vmware.guestInfo [$VSPHERE_VAPP_TRANSPORT]
   --vmwarevsphere-vcenter                                                                                              vSphere IP/hostname for vCenter [$VSPHERE_VCENTER]
   --vmwarevsphere-vcenter-port "443"                                                                                   vSphere Port for vCenter [$VSPHERE_VCENTER_PORT]
flag provided but not defined: -vmwarevsphere-graceful-shutdown-timeout
```

Later from we also get like this:
```
fleet-default/alpha-lab10321-worker1-e88fa4ef-ggdln-machine-provision-slls9[machine]: Machine config for alpha-lab10321-worker1-e88fa4ef-ggdln does not exists, so nothing to do...
fleet-default/alpha-lab10321-worker1-e88fa4ef-ggdln-machine-provision-slls9[machine]: About to remove alpha-lab10321-worker1-e88fa4ef-ggdln
```

Going to change back to 105 and verify that it's still working.

**Comment by akpn** [comment]
I work with @sennerholm and have been doing some additional testing with 2.7.7
Using a fresh cluster managed by a Rancher deployment upgraded from 2.7.6 to 2.7.7 and setting:
```
- name: CATTLE_MACHINE_PROVISION_IMAGE
  value: 'rancher/machine:v0.15.0-rancher100'
```
After which I scaled the cluster from 6 to 12 worker nodes - all provisioning jobs were successful except for 1 giving me a CreateError and I could view the following in the logs:
```
2023-10-02 09:09:28.873 | flag provided but not defined: -vmwarevsphere-provider-id
..
2023/10/02 07:09:41 [INFO] [machineprovision] fleet-default/alpha-lab10262-worker2-c8933d60-6fh5c: Failed to create infrastructure for machine alpha-lab10262-worker2-55db467554-lncrn, deleting and recreating...
```
The re-tried provisioning job succeded
I am currently running additional scale operations to see if this error keeps appearing

**Comment by akpn** [comment]
After running scale down operations on the same cluster I started getting the same error as @sennerholm 
```
_time: 2023-10-02 13:16:13.020
message: flag provided but not defined: -vmwarevsphere-graceful-shutdown-timeout
kubernetes.container_image: docker.io/rancher/machine:v0.15.0-rancher100
```
It seems like this flag from https://github.com/rancher/rancher/issues/42308 has started appearing in some config/template used in 2.7.7

**Comment by snasovich** [comment]
@sennerholm @akpn , thank you again for detailed reports. I see what's happening - the version of `rancher/machine` shipped with Rancher release is the one that is executed to populate CRDs like `vmwarevsphereconfigs.rke-machine-config.cattle.io` and since it's version `v0.15.0-rancher104` it already has that field. So then when RKE2 provisioning is overridden to use `v0.15.0-rancher100` it doesn't have that new `vmwarevsphere-graceful-shutdown-timeout` field and fails.

TLDR: For vSphere provisioning, the workaround is to use version `v0.15.0-rancher105` as per your previous [comment](https://github.com/rancher/rancher/issues/42972#issuecomment-1741100397).
Please confirm it's working for you on `v0.15.0-rancher105` and whether you're getting any errors/warnings?

**Comment by sennerholm** [comment]
@snasovich I understand it as the way forward is to release a 2.7.8 with some fixes in, and even if we can override the environment on the first install in our production environment to point to `0.15.0-rancher100` we can't assume that customers skip to install 2.7.7 on it's way to 2.7.8, so we really need to go forward to newer version of rancher/machine (`0.15.0-rancher105`). We try to test to set the machine environment to `0.15.0-rancher105` and scale up and down our noodpools (replaces members) on VMware/vSphere. Anything else you want us to test? 

We really want to have a stable 2.7.7 or newer out, because we are impacted by a clusterapi race condition, that seems to be fixed in 2.7.7

**Comment by snasovich** [comment]
@sennerholm , those are valid concerns. We're going to test how "downgrade" scenario works if we reference `0.15.0-rancher100` in 2.7.8 RCs when upgrading from 2.7.7 that references `0.15.0-rancher104` - I _think_ the CRD will get updated to have the new field removed but I'm not sure yet. There may also be some caveats with dynamic schema as preserved on machine pools created with 2.7.7 that references `0.15.0-rancher104`/`0.15.0-rancher105`.

**Comment by sennerholm** [comment]
Downgrading to `0.15.0-rancher100` doesn't seems to remove the field for us, so I think we need to forward to '0.15.0-rancher105'

**Comment by sennerholm** [comment]
@snasovich, thanks for the explanation. 
Only RKE2, but then I have a better understanding on why the CRDs wasn't updated, if you create an RC I can try to test it on my side, we have some automation to spawn clusters on VMware. What we try after is to go from:
 nr of nodes in etcd worker1 worker2: 3 3 3 -> 5 10 3 -> 3 3 3 -> 3 10 3 -> 3 3 3. 
 I'm at the 3 10 3 stage and green so far with 105.

**Comment by snasovich** [comment]
@sennerholm , https://github.com/rancher/rancher/releases/tag/v2.7.8-rc1 is now available and it is based on the approach where `rancher/machine` is downgraded to version `0.15.0-rancher100`.
This should remove the new `vmwarevsphere-graceful-shutdown-timeout` field from `vmwarevsphereconfigs.rke-machine-config.cattle.io` CRD upon upgrading.
Note that if you've applied a workaround to 2.7.7 where `rancher/machine` was bumped to version 105 via `CATTLE_MACHINE_PROVISION_IMAGE` env var and successfully provisioned vSphere RKE2 clusters, the nodes in these clusters' machine pools will likely reprovision on upgrade to 2.7.8(-rc1). We're going to test that too.

**Comment by sennerholm** [comment]
@snasovich It seems to work as you said.
The field is removed, the cluster nodes are reprovisioned. 

I will try to test an 2.7.6->2.7.8-rc1 upgrade also to check if the nodes are reprovisioned or not.

**Comment by bfbachmann** [comment]
Reopening this issue as there is a lingering bug that I'm addressing [here](https://github.com/rancher/machine/pull/228).

The result of the bug is that, if you create a vSphere cluster and then rotate your vSphere username, the Rancher machine will continue using your old username instead of the new one.

**Comment by slickwarren** [comment]
is this affecting 2.7.8 as well @bfbachmann ?

**Comment by bfbachmann** [comment]
Closing this issue as it should not have been reopened in the first place. https://github.com/rancher/rancher/issues/40608 Is the issue that should have been reopened instead.
